tests/snap_test.go

```diff
@@ -0,0 +1,153 @@
+	}
+	if err := os.Remove("./chip-tool.log"); err != nil && !os.IsNotExist(err) {
+		t.Fatalf("Error deleting log file: %s\n", err)
+	}
+
+	// Setup: run and log chip-all-clusters-minimal-app in the background
+	logFile, err := os.Create(chipAllClusterMinimalAppLog)
+	if err != nil {
+		t.Fatalf("Error creating log file: %s\n", err)
+	}
+
+	cmd := exec.Command("./" + chipAllClusterMinimalAppFile)
+	cmd.Stdout = logFile
+	cmd.Stderr = logFile
+
+	err = cmd.Start()
+	if err != nil {
+		t.Fatalf("Error starting application: %s\n", err)
+	}
+
+	t.Cleanup(func() {
+		matches, err := filepath.Glob("/tmp/chip_*")
+		if err != nil {
+			t.Fatalf("Error finding tmp chip files: %s\n", err)
+		}
+
+		for _, match := range matches {
+			if err := os.Remove(match); err != nil {
+				t.Fatalf("Error removing tmp chip file %s: %s\n", match, err)
+			}
+		}
+
+		if err := cmd.Process.Kill(); err != nil {
+			t.Fatalf("Error killing process: %s\n", err)
+		}
+
+		if logFile != nil {
+			logFile.Close()
+		}
+	})
+
+	t.Run("Commission", func(t *testing.T) {
+		utils.Exec(t, "sudo chip-tool pairing onnetwork 110 20202021")
+	})
+
+	t.Run("Control", func(t *testing.T) {
+		utils.Exec(t, "sudo chip-tool onoff toggle 110 1")
+		waitForAppMessage(t, "./"+chipAllClusterMinimalAppLog, "CHIP:ZCL: Toggle ep1 on/off", start)
+	})
+}
+
+func setup() (teardown func(), err error) {
+	const chipToolSnap = "chip-tool"
+
+	log.Println("[CLEAN]")
+	utils.SnapRemove(nil, chipToolSnap)
+
+	log.Println("[SETUP]")
+
+	teardown = func() {
+		log.Println("[TEARDOWN]")
+		utils.SnapDumpLogs(nil, start, chipToolSnap)
+
+		log.Println("Removing installed snap:", !utils.SkipTeardownRemoval)
+		if !utils.SkipTeardownRemoval {
+			utils.SnapRemove(nil, chipToolSnap)
+		}
+	}
+
+	if utils.LocalServiceSnap() {
+		err = utils.SnapInstallFromFile(nil, utils.LocalServiceSnapPath)
+	} else {
+		err = utils.SnapInstallFromStore(nil, chipToolSnap, utils.ServiceChannel)
+	}
+	if err != nil {
+		teardown()
+		return
+	}
+
+	// connect interfaces
+	utils.SnapConnect(nil, chipToolSnap+":avahi-observe", "")
+	utils.SnapConnect(nil, chipToolSnap+":bluez", "")
+	utils.SnapConnect(nil, chipToolSnap+":process-control", "")
+
+	return
+}
+
+func waitForAppMessage(t *testing.T, appLogPath, expectedLog string, since time.Time) {
+	const maxRetry = 10
+
+	for i := 1; i <= maxRetry; i++ {
+		time.Sleep(1 * time.Second)
+		t.Logf("Retry %d/%d: Waiting for expected content in logs: %s", i, maxRetry, expectedLog)
+
+		logs, err := readLogFile(appLogPath)
+		if err != nil {
+			t.Fatalf("Error reading log file: %s\n", err)
+			continue
+		}
+
+		if strings.Contains(logs, expectedLog) {
+			t.Logf("Found expected content in logs: %s", expectedLog)
+			return
+		}
+	}
+
+	t.Fatalf("Time out: reached max %d retries.", maxRetry)
+}
+
+func readLogFile(filePath string) (string, error) {
+	text, err := os.ReadFile(filePath)
+	if err != nil {
+		return "", err
+	}
+	return string(text), nil
+}
```

Comment:
This could be marked as a helper function:

```go 
func readLogFile(t *testing.T, filePath string) (string, error) {
    t.Helper()
    text, err := os.ReadFile(filePath)
    if err != nil {
        return "", err
    }
    return string(text), nil
}
```
(then change the in the call of the function also to: `logs, err := readLogFile(t, appLogPath)` )

Reply:
Yes, this can be refactored to not only be flagged as a test helper, but also to handle errors internally instead of returning to the caller.
At this state, the function doesn't do much and `os.ReadFile` can be simply called by the caller.

Note that the testing code isn't written as part of this PR. It is just moved from another repo; see first commit: https://github.com/canonical/chip-tool-snap/pull/36/commits/57cfa2264ecb7725ac4e530b6d66b2e2f98d6465

Reply:
Created an issue: https://github.com/canonical/chip-tool-snap/issues/39

Reply:
Oh, I see, thank you for letting me know @farshidtz!